### PR TITLE
fix(rule): allow proven one-shot RTL createRef components

### PR DESCRIPTION
**Pull request by aidenybai**
## Why

`no-create-ref-in-function-component` reports an authentic test-local helper in `vrk-kpa/suomifi-ui-components@b3b68b8cd68127b7828300eb790ae610253f5557` (`src/core/Modal/Modal.test.tsx:174`). The helper creates a ref only so React Testing Library can mount one isolated component root once. It is never rerendered, escaped, or reused, so ref identity cannot be observed across renders.

The precision boundary is semantic and intentionally narrow. React `StrictMode` may invoke a discarded render, but it does not preserve the first ref as component state; separate `render()` calls create independent roots. Both are therefore clean. Reusing the same root through `rerender`, allowing the component or result to escape, using an unproven render function, or adding observable render-time behavior remains reportable.

## What changed

- Proves exact React Testing Library named/namespace import provenance and test-callback ownership before allowing the one-shot case.
- Requires a direct local component, identifier-only component parameters, local `createRef` declarations, a tightly bounded JSX return, and independent direct RTL roots whose results cannot rerender or escape.
- Accepts renamed plain render-result bindings while rejecting executable destructuring defaults, parameter defaults/patterns/rest parameters, same-root rerenders, render options, loops/conditionals, aliases, helpers, custom render APIs, component/ref escapes, nested callbacks, state/effects/context/hooks, and product code.
- Adds paired authentic and adversarial regressions, permanent false-positive and true-positive fuzz seeds, a generator shape, and a patch changeset.

## Real-repository validation

Current comparison: baseline `b2f3c9c463f5f04dd6d7c2d09621ff9edcc7144f`, candidate `dd28938693c989af384c3907bae0c2ff27d22b67`.

Pinned Suomi scan using built baseline and candidate CLIs:

| Revision | Files scanned | Target findings | Delta |
| --- | ---: | ---: | --- |
| baseline | 325/325 | 2 | — |
| candidate | 325/325 | 1 | removed only `Modal.test.tsx:174` |

The real product finding at `src/core/Table/Table.tsx:195` remains. There were no added target diagnostics or scan failures.

An additional pinned Suomi probe wrapped the authentic modal JSX argument in `as React.ReactElement`. Baseline still reported both findings; the current candidate removed only the wrapped modal test finding and preserved the Table finding, with 325/325 files completed in both scans.

Across 17 React Bench solution/oracle trees, baseline and candidate produced the same 6 target diagnostics over 76 source files containing `createRef`. Fourteen files were unsupported by the direct AST scanner in both builds; manual inspection confirmed they are class-component constructor usage or test-callback refs outside this exemption, so they are recorded as unsupported evidence rather than scanner passes.

### Local OSS corpus

The isolated schema-v3-compatible RDE harness compared baseline `b2f3c9c463f5f04dd6d7c2d09621ff9edcc7144f` with candidate `f959a1ab5b4a07a401270c26519fb325b6bab280` over the same frozen first 100 requested roots with fresh caches and serial execution:

| Evidence | Baseline | Candidate | Result |
| --- | ---: | ---: | --- |
| Completed by the bulk harness | 99 | 99 | identical repository/ref/root order |
| Target findings across completed roots | 3 | 3 | 0 added, 0 removed |
| Full diagnostic hashes across completed roots | — | — | identical for all 99 |

The remaining pinned `excalidraw/excalidraw@0457ac90634b8556feaa9c0e56be94d3f13fd9bd` repo-root scan exceeded the full harness's ten-minute limit in both builds. It was not treated as a pass. A supplemental target-rule-only A/B disabled unrelated analyzers, used fresh caches and an 8 GB heap, and completed all 4 discovered projects and 467/467 files in each build: 0 target findings in both, no skipped or incomplete projects. This supplies target-rule evidence for the 100th requested root without hiding the shared full-scan timeout.

The follow-up `dd2893869` review fix is a monotonic expansion through the existing transparent-expression-root utility: it recognizes only parentheses and TypeScript `as`, `satisfies`, and non-null wrappers around an already-proven direct JSX argument. Nontransparent sequences, conditionals, arrays, calls, aliases, render options, and rerender paths remain blocked. The exact Suomi wrapper probe, focused paired tests, permanent fuzz seed, strict generated fuzz, and React Bench fuzz were rerun on that follow-up; the frozen bulk RDE verdicts above are retained as the pre-expansion full-diagnostic comparison.

A final production target-only RDE supplement compared exact baseline `b2f3c9c463f5f04dd6d7c2d09621ff9edcc7144f` with exact candidate `dd28938693c989af384c3907bae0c2ff27d22b67` across all 100 frozen roots. Both completed 100/100 roots and 25,758/25,758 analyzed files with no scan failures, skipped checks, or non-React roots. Target findings stayed 3 → 3 with no added or removed rule/file/line/column verdicts.

## Test plan

- Focused one-shot regressions — 22/22 passed
- `nr --filter oxlint-plugin-react-doctor test` — 568 files passed; 14,815 tests passed; 188 skipped
- `FUZZ_RULE=no-create-ref-in-function-component FUZZ_STRICT=1 FUZZ_ITERATIONS=500 nr fuzz` — passed; target fired
- The same strict 500-iteration run against `/Users/aidenybai/Developer/react-bench-5` — passed; target fired
- `nr --filter @react-doctor/fuzz test` — passed
- `nr test` — 14/14 tasks; 2,245 tests passed; 27 skipped
- `nr typecheck` — passed
- `nr lint` — passed with only pre-existing warnings
- `nr format:check` — passed
- `git diff --check` — passed
- Post-change `truffler` searches — no duplicate helper found

---

> [!NOTE]
> **Medium Risk**
> The change is a targeted lint exemption with many adversarial tests, but incorrect proof logic could hide real `createRef` bugs in tests or miss edge cases in RTL mount patterns.
> 
> **Overview**
> Fixes **`no-create-ref-in-function-component`** false positives for test-local components that only mount once via React Testing Library.
> 
> Adds **`isProvenOneShotTestingLibraryComponent`**, which skips reporting when a function in a test-like file is a `const` inside an `it`/`test` callback, its body is only `createRef` declarations plus simple JSX, and every reference is a direct `render(<Component />)` from `@testing-library/react` with no `rerender`, result escape, aliases, or risky render options. Transparent TS wrappers (`as`, `satisfies`, `!`, parens) on the JSX argument are allowed; rerenders, hooks/state in the helper, ref escapes, and production code still report.
> 
> **`no-create-ref-in-function-component`** calls this helper before existing write-only checks. Regression tests, fuzz corpus seeds, snippet pools, and a patch changeset cover allowed and blocked shapes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit dd28938693c989af384c3907bae0c2ff27d22b67. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by aidenybai** [comment]
Implementation is intentionally blocked after the same-file soundness audit.

The exact candidate file has one visible `render(<ModalWithFocusOnOpenRef />)` call and no local rerender, export, or `StrictMode`. That is enough to establish the authentic Suomi behavior, but it is not enough for a general per-file exemption:

```ts
// jest.setup.ts — outside the file oxlint is visiting
import { configure } from "@testing-library/react";

configure({ reactStrictMode: true });
```

Suomi pins `@testing-library/react@16.0.1`. At that exact upstream tag (`3dcd8a9649e25054c0e650d95fca2317b7008576`), `src/pure.js` reads this global configuration and wraps every later `render(ui)` in `React.StrictMode`. Its own `src/__tests__/render.js` asserts that a single `render(<Component />)` invokes the component twice when the flag is enabled.

Therefore these two projects present byte-identical AST and filename input to this rule:

1. Suomi: no global configuration, one render.
2. Another project: setup-file `configure({ reactStrictMode: true })`, two renders.

The current oxlint rule receives only the current file's AST/scopes/CFG; it cannot inspect Jest setup files or Testing Library's process-global configuration. Checking for local `StrictMode`, local `rerender`, one JSX reference, or no export would suppress the second project's positive too.

No name/path exception or incomplete “one mount” proof will be implemented. This draft stays regression-only until React Doctor has project-level test-runtime capability analysis or a sound host-provided signal.

**Comment by aidenybai** [comment]
Superseding the earlier blocker analysis: React Testing Library's global `reactStrictMode` configuration is not a positive for this rule. Strict Mode may invoke a discarded render, but the first `createRef()` result is not retained as component state or observed by the committed tree. Likewise, two independent `render()` roots do not require stable ref identity across renders.

The pushed implementation therefore treats binding-proven Strict Mode and multiple independent roots as clean, while preserving diagnostics for same-root `rerender`, escaped results/components/refs, unproven render APIs, options, loops/conditionals, and observable render behavior. The updated PR body contains the full pinned A/B and fuzz evidence.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/is-proven-one-shot-testing-library-component.ts [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+  if (
+    !callExpression ||
+    !isNodeOfType(callExpression, "CallExpression") ||
+    callExpression.arguments.length !== 1 ||
+    callExpression.arguments[0] !== rootElement ||
+    !isProvenTestingLibraryRenderCall(callExpression, scopes)
+  ) {
+    return null;
+  }
+  return callExpression;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/is-proven-one-shot-testing-library-component.ts:332]: Fixed in dd2893869. The proof now climbs only through the existing transparent-expression wrapper set before requiring that the resulting node is the sole direct RTL render argument. Added clean regressions for parentheses, `as`, `satisfies`, and TS non-null wrappers; sequence, conditional, array, call-wrapper, and JSX-alias near-neighbors still report. Focused 22/22, full plugin 14,815 passed, strict generated + React Bench fuzz passed, and the exact pinned Suomi wrapper A/B removed only the Modal test FP while preserving the Table product finding (325/325 files in both).
